storageclass ceph add imageformat parameter

weiwei04 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#pull-request-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
Add a imageformat parameter for StorageClass(ceph rbd)

k8s hard coded ceph imageformat 1, according to [ceph manual](http://docs.ceph.com/docs/master/man/8/rbd/), imageformat 1 was deprecated, we should add an extra ceph parameter to set ceph rbd imageformat. Ceph rbd imageformat can only be 1 or 2, set the default value to 1.

**Release note**:
```release-note
Allow StorageClass Ceph RBD to specify image format and image features.
```

weiwei04 committed:
storageclass ceph add imageformat parameter

weiwei04 commented:
/assign @saad-ali please take a look.

weiwei04 commented:
@saad-ali any comments on this pr?

weiwei04 commented:
/assign @rootfs would you please take a look, thanks.

rootfs commented on pkg/volume/rbd/rbd.go:
```diff
@@ -281,11 +284,17 @@ func (r *rbdVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
 	}
 	// sanity check
+	if imageFormat != rbdImageFormat1 && imageFormat != rbdImageFormat2 {
+		return nil, fmt.Errorf("invalid ceph imageformat %s, neither 1 or 2", imageFormat)
```
fmt.Error("invalid ..., expecting %s or %s", imageFormat1, imageFormat2)

rootfs commented:
@k8s-bot ok to test

rootfs commented:
Though image format 1 is deprecated, it is still used by krbd (which rbd volume is based on). Newer kernel starts to support format 2 but many linux distros don't have them yet. The plan is to wait till they pick up the newer kernel and we add image format options.

weiwei04 commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

weiwei04 commented:
@rootfs modified the error log, please take a look, thanks.

rootfs commented:
@leseb

rootfs commented:
@weiwei04 please update readme and more diagnostic messages when older krbd failes to map v2 image

Commit:
Add `imageFeatures` parameter for RBD volume plugin, which is used to
customize RBD image format 2 features.
Update RBD docs in examples/persistent-volume-provisioning/README.md.

cofyc commented:
@k8s-ci-robot CLA signed.

cofyc commented:
Bump to verify CLA.

rootfs commented on examples/persistent-volume-provisioning/README.md:
```diff
@@ -192,6 +193,10 @@ parameters:
 * `userId`: Ceph client ID that is used to map the RBD image. Default is the same as `adminId`.
 * `userSecretName`: The name of Ceph Secret for `userId` to map RBD image. It must exist in the same namespace as PVCs. It is required.
+* `imageFormat`: Ceph RBD image format, "1" or "2". Default is "1".
+* `imageFeatures`: Ceph RBD image format 2 features, comma delimited. This is optional, and only be used if you set `imageFormat` to "2". For a complete list of available image features, please refer to [RBD docs](http://docs.ceph.com/docs/master/man/8/rbd/). By default, all features (except for striping) will be enabled.
```
Not all features are supported and unfortunately RBD docs are not well updated.

rootfs commented:
@weiwei04 @cofyc The readme and feature validation need some work, see my closed PR here
https://github.com/kubernetes/kubernetes/pull/43816

We cannot turn on exclusive lock feature for now, exclusive lock and advisory lock cannot work together.  For now, just turn on layering feature and we'll revisit exclusive lock later.

rootfs commented:
/assign

Commit:
Only `layering` RBD image format 2 feature should be supported for now.

rootfs commented:
@leseb want to take a look?

cofyc commented:
@rootfs 

Thanks for reviewing this.
 
A slight difference with your PR https://github.com/kubernetes/kubernetes/pull/43816, we use two parameters `imageFormat` and `imageFeatures` instead one `features` parameter, but I think it's better to be consistent with [RBD options](docs.ceph.com/docs/master/man/8/rbd/) which have two options to configure these. And it makes it possible to use image format 2 without any features enabled.

weiwei04 commented:
/release-note-none

weiwei04 commented:
/release-note

leseb reviewed (approved):
@rootfs lgtm

cofyc commented:
@rootfs 
Is there anything else I need to fix? Are you ok with RBD new parameters, and theirs default behaviors?

rootfs commented:
/lgtm

thanks for contributing!

rootfs commented:
/approve

eparis commented:
/approve

ianchakeres commented:
/sig storage

ianchakeres commented on examples/persistent-volume-provisioning/README.md:
```diff
@@ -209,6 +209,7 @@ parameters:
     pool: kube
     userId: kube
     userSecretName: ceph-secret-user
+    imageFormat: "1"
```
Does "1" need to be in double quotes?

cofyc replied:
Yes, it need to be in double quotes, because `parameters` is a map of string to string.

ianchakeres commented:
Will there ever be a case when imageFormat will be nil or ""? I'm not sure exactly where in the lifecycle of a PV Provision() is called.

For example, if a PV were created before this code was merged.

cofyc commented:
@ianchakeres 

hi,

`imageFormat` is type `string`, so it will never to `nil`. And it will never be empty string (""), because `imageFormat` is invalid option before this code is merged.

After this code is merged, `imageFormat` can be missing (not configured) or one of possible values ("1" or "2"). If `imageFormat` is configured, Provisioner will use configured value to create PVs. If `imageFormat` is missing, Provisioner will use default value ("1") to create PVs, so old RBD StorageClasses which do not (and can not) specify `imageFormat` parameter will continue work as before.
